### PR TITLE
Execution comparison

### PR DESCRIPTION
**Pull request by farajidaneshgar**
execution comparison:
-Introduce a module to compare two groups of executions using a differential flame graph.
-An "EventDensityView" is used to select a time range for each group and Select the desired traces in an experiment.
-Showing the difference ratio based on self time required a modification to AggregatedCalledFunction.
-Introducing "ParametricWeightedTreeUtils" to get "SelfTime" instead of "duration"  while building the differential Trees.
-Adding the option of textual input of dates and query to build the differential flame graph.
-moving grouping to hamburger menu
-Differential flame graph view is updated to be able to show and hide Query part by demand.
    
remove commented code: Done
improve commit messages, no "second commit" and "third commit":Done
fix dependencies:Done
update aggregatecallsite to have functions like getWeight(string). This doesn't compile locally:Done

Dependencies to the incubator callstack still exist and need to be removed: Done
Modifications to the incubator callstack need to be moved to trace compass core profiling plugin.:Done

### CODE REVIEW

**Comment by MatthewKhouzam** [comment]
Can you make the build pass?

**Comment by MatthewKhouzam** [comment]
I tried using it yesterday.

1. The code doesn't build on HEAD. This is because Arnaud deprecated the incubator callstack
2. There is open tracing changes in the first patch, they should be a separate patch
3. splitting the code into "first patch" "second patch" and "third patch" is not descriptive enough. We need to have titles that describe what they do.

I have a use case that can use it. I can share the traces, but I cannot even get it to work at the moment!

**Comment by farajidaneshgar** [comment]
Hello Matthew,
Yes. I have new push eliminating that dependencies and adding some other
features. I will pull it today.

**Comment by arfio** [comment]
Just writing this comment to update the situation:
- Dependencies to the incubator callstack still exist and need to be removed
- Modifications to the incubator callstack need to be moved to trace compass core profiling plugin.

**Comment by MatthewKhouzam** [comment]
please update commit messages, also thanks for fixing the pr!

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ org.eclipse.core.runtime,
+ org.eclipse.tracecompass.common.core,
+ org.eclipse.tracecompass.tmf.core,
+ org.eclipse.tracecompass.tmf.ui,
+ org.eclipse.tracecompass.incubator.callstack.ui,
+ org.eclipse.tracecompass.incubator.callstack.core,
+ org.eclipse.tracecompass.incubator.analysis.core,
+ org.eclipse.tracecompass.incubator.executioncomparision.core,
+ org.eclipse.jdt.annotation;bundle-version="[2.0.0,3.0.0)";resolution:=optional,
+ jakarta.validation.jakarta.validation-api,
```
Comment (MatthewKhouzam) [inline analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF:19]: remove

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ org.eclipse.tracecompass.analysis.profiling.core
```
Comment (MatthewKhouzam) [inline analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF:20]: put above annotations

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java [hunk 3]
```diff
@@ -192,7 +181,7 @@ public IWeightedTreeSet<ICallStackSymbol, Object, DifferentialWeightedTree<ICall
      *            the monitor, can be null
      * @return the differential weighted provider or null
      */
-    public @Nullable DifferentialWeightedTreeProvider<?> getDiffProvider(@Nullable IProgressMonitor monitor) {
+   public @Nullable DifferentialWeightedTreeProvider<?> getDiffProvider(@Nullable IProgressMonitor monitor) {
```
Comment (MatthewKhouzam) [inline analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java:195]: revert

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java [hunk 4]
```diff
@@ -360,7 +368,16 @@ public void dispose() {
      */
     private void setDifferentialCallGraphProvider(DifferentialCallGraphProvider differentialCallGraphProvider) {
         fDifferentialCallGraphProvider = Objects.requireNonNull(differentialCallGraphProvider);
-
+    }
+    private static ITmfTrace getTrace(String traceName) {
```
Comment (MatthewKhouzam) [inline analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java:372]: this needs to be annotated <at>nullable guess what, in github, nullable is a user that gets pinged!! @PatrickTasse

**Review by MatthewKhouzam — CHANGES_REQUESTED** [review]
Some first pass changes. you need to fix imports too. I am trying to get this working too!

**Comment by ebugden** [comment]
Hello! Are there screenshots or a video screen capture that could be used to give feedback on the result without needing to do the whole development setup? My understanding is that building it myself is currently the only way.

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ org.eclipse.tracecompass.incubator.analysis.core,
+ org.eclipse.tracecompass.analysis.timing.core,
+ org.eclipse.tracecompass.tmf.core,
+ org.eclipse.tracecompass.tmf.analysis.xml.core,
+ org.eclipse.tracecompass.tmf.ctf.core,
+ org.eclipse.tracecompass.lttng2.common.core,
+ org.eclipse.tracecompass.lttng2.ust.core,
+ org.eclipse.tracecompass.analysis.os.linux.core,
+ org.eclipse.tracecompass.analysis.profiling.core,
+ org.eclipse.tracecompass.incubator.callstack.core,
```
Comment (arfio) [inline analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/META-INF/MANIFEST.MF:25]: This line should be removed.

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -0,0 +1,34 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: %Bundle-Name
+Bundle-Vendor: %Bundle-Vendor
+Bundle-SymbolicName: org.eclipse.tracecompass.incubator.executioncomparision.core;singleton:=true
```
Comment (arfio) [inline analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/META-INF/MANIFEST.MF:5]: Ideally, a refactor to rename this package to executioncomparison should be done.

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialCallGraphProvider.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
+/*******************************************************************************
+ * Copyright (c) 2023 École Polytechnique de Montréal
```
Comment (arfio) [inline analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialCallGraphProvider.java:2]: All of the copyrights may be updated to 2024

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/Messages.java
Comment (arfio) [inline analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/Messages.java:1]: Is this file needed ?

File: callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AbstractCalledFunction.java
Comment (arfio) [inline callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AbstractCalledFunction.java:1]: This change should be removed from this PR and moved to another if needed.

File: callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AggregatedCalledFunction.java
Comment (arfio) [inline callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AggregatedCalledFunction.java:1]: This change should be removed from this PR and moved to another if needed.

**Review by arfio — CHANGES_REQUESTED** [review]

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+     * @return WeightedTreeSet<ICallStackSymbol, Object> the merged call graph
+     *
+     */
+    public WeightedTreeSet<ICallStackSymbol, Object> mergeCallGraph(ITmfTimestamp start, ITmfTimestamp end, List<String> traceList) {
+        try (ScopeLog sl = new ScopeLog(LOGGER, Level.FINE, "DifferenentialSequenceCGA::MergeCallGraph")) { //$NON-NLS-1$
+            List<CallGraph> cGList = new ArrayList<>();
+            WeightedTreeSet<ICallStackSymbol, Object> newTreeSet = new WeightedTreeSet<>();
+            String mainGroup = MERGE;
+
+            for (String traceName : traceList) {
```
[the remark this answers is not in the document]
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java:147]: this can't be null. In initialization step (    private List<String> fTraceListA = new ArrayList<>(); private List<String> fTraceListB = new ArrayList<>();) traceLists are initialized as a empty list. And they will be filled by the signal handling before calling this function. However if it is null, the function returns an empty weightedTreeSet.

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF (hunk 1 above)
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF:19]: Done

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF (hunk 2 above)
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF:20]: Done

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java (hunk 4 above)
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java:372]: I can't see why, its called in this loop: for (String traceName : traceList) { ITmfTrace trace = getTrace(traceName);} and it can not be null.

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/META-INF/MANIFEST.MF (hunk 5 above)
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/META-INF/MANIFEST.MF:25]: Done

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/META-INF/MANIFEST.MF (hunk 6 above)
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/META-INF/MANIFEST.MF:5]: Done

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialCallGraphProvider.java (hunk 7 above)
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialCallGraphProvider.java:2]: Done

File: callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AbstractCalledFunction.java
Comment (farajidaneshgar) [reply callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AbstractCalledFunction.java:1]: Done

File: callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AggregatedCalledFunction.java
Comment (farajidaneshgar) [reply callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AggregatedCalledFunction.java:1]: Done

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core.tests/META-INF/MANIFEST.MF [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+Bundle-Name: %Bundle-Name
+Bundle-Vendor: %Bundle-Vendor
+Bundle-SymbolicName: org.eclipse.tracecompass.incubator.executioncomparision.core.tests
+Bundle-Version: 0.1.0.qualifier
+Bundle-Localization: plugin
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Require-Bundle: org.eclipse.core.runtime,
+ org.eclipse.core.resources,
+ org.eclipse.tracecompass.common.core,
+ org.eclipse.tracecompass.incubator.executioncomparision.core,
```
[the remark this answers is not in the document]
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.core.tests/META-INF/MANIFEST.MF:12]: why should it be removed? It s needed in Activator class of test package.

File: analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+    private static final String MERGE = "Merge"; //$NON-NLS-1$
+    private @Nullable DifferentialCallGraphProvider fDifferentialCallGraphProvider;
+    private static Map<String, String> fcallStackAnalysisMap = new HashMap<>();
+    private static Map<String, ICallGraphProvider2> fTraceCallGraphRegistry = new HashMap<>();
+    private ITmfTimestamp fStartA = TmfTimestamp.BIG_BANG;
+    private ITmfTimestamp fEndA = TmfTimestamp.BIG_CRUNCH;
+    private ITmfTimestamp fStartB = TmfTimestamp.BIG_BANG;
+    private ITmfTimestamp fEndB = TmfTimestamp.BIG_CRUNCH;
+    private String fStatistic = ""; //$NON-NLS-1$
+    private List<String> fTraceListA = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (farajidaneshgar) [reply analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparision/core/DifferentialSeqCallGraphAnalysis.java:85]: Done

**Comment by farajidaneshgar** [comment]
Hi ebugden, here is some screenshots:
Screenshot from 2024-05-16 12-11-56

**Comment by farajidaneshgar** [comment]
Screenshot from 2024-05-16 12-12-36

Screenshot from 2024-05-16 12-12-27

Screenshot from 2024-05-16 12-12-22

Screenshot from 2024-05-16 12-12-17

Screenshot from 2024-05-16 12-12-11

Screenshot from 2024-05-16 12-12-06

**Comment by ebugden** [comment]
Thank you for the screenshots and for your work on this analysis! :sunflower: It's practical to be able to see the result visually in order to give feedback.

Some elements I like:

* The [diverging colour scale](https://blog.datawrapper.de/wp-content/uploads/2021/03/full-210101_colorscales25-1024x367.png)!  The encoding choice of blue (for better) and red (for worse) is also good for folks who are red/green colour blind.
* I think the percentage at the front of the label makes it easier to notice that that's the primary information of the visualisation.
* The "Trace name" label (above the trace selection area).

## Suggestions

I have some questions and comments about the visualisation.

### Clarify how to read the chart

I noticed that there is a label of "GroupB-GroupA" next to the chart, but it's still not clear what data the chart contains. Some questions that came to mind were:

* What is being "diff"-ed between Group B and Group A? Since it's a flame graph, is it the difference in percentage of execution time? Is it the difference in execution time?
* Which proportions (the rectangles labelled with function names) are displayed in the flame graph? Is it the proportions (distribution of execution time) of Group B? Is it the proportions of Group B that have comparable proportions in Group A?
  * How are functions in Group B that are not in Group A displayed? How are functions that are in Group A, but not in Group B displayed?

#### Suggestions

* I think there should be clear documentation somewhere about what data the chart shows and how to read it that answers these questions.
* Clarify that the visualisation is a kind of Flame Graph. I think for people who are just discovering this visualisation it would not be possible to know. E.g. Put a label "Flame Graph" somewhere.
* Clarify the structure of the visualisation (i.e. that the top section is just data selection, the bottom section is the visualisation). This could be done with labels, but I think it's possible that applying some of the suggested changes below would improve this (e.g. making the histogram data in the graphical selection area all the same colour)

sequential-trace-comparison-annotated-sections

### Clarify when this chart should be used

At the moment it's unclear to me what kind of things it is appropriate to compare with this visualisation. I feel that answering the above questions (in the "Clarify how to read the chart" section) would improve this.

### Miscellaneous suggestions

I included screenshots annotated with my comments below.

Some important themes:

* Clarify axis titles and labels (and remove when not relevant). This will help with clarifying what is being visualised both in the graphical selection and in the main chart.
* Make histogram data all the same colour (in the graphical selection area)

sequential-trace-comparison-annotated

select-relevant-traces-annotated

graphical-select-annotated

**Comment by ebugden** [comment]
Adding comments from an offline discussion.

### How to read the chart

- **Structure:** The structure of the Flame Graph is defined exclusively by the Group B selection (i.e. the proportions and names are defined by Group B)
- **Colour:** The colour of the bars is determined by the difference in absolute execution time between Group B and Group A which is then encoded by a divergent (classified?) colour scale (blue: less time than A, red: more time than A).

#### Comments

- I think how the chart is read needs to be further clarified in the chart itself (e.g. clearer labels, legend, textual description)
- Arguably proportions are the most important information. However with the current implementation, the longest portions of execution are far to the right and also there is a lot of visual emphasis on work done by B that is not done by A because they show up as dark red and because the largest (which isn't necessarily the most important information).

**Comment by MatthewKhouzam** [comment]
This was merged! :) :) :) :+1: :100:
